contrib/lib/src/compression/mod.rs

```diff
@@ -0,0 +1,138 @@
+//! `Compression` fairing and `Compressed` responder to automatically and
+//! on demand respectively compressing responses.
+mod context;
+mod fairing;
+mod responder;
+
+pub use self::fairing::Compression;
+pub use self::responder::Compressed;
+
+crate use self::context::Context;
+use rocket::http::hyper::header::{ContentEncoding, Encoding};
+use rocket::http::uncased::UncasedStr;
+use rocket::http::MediaType;
+use rocket::{Request, Response};
+use std::io::Read;
+
+#[cfg(feature = "brotli_compression")]
+use brotli::enc::backward_references::BrotliEncoderMode;
+
+#[cfg(feature = "gzip_compression")]
+use flate2::read::GzEncoder;
+
+crate struct CompressionUtils;
+
+impl CompressionUtils {
+    fn accepts_encoding(request: &Request, encoding: &str) -> bool {
+        request
+            .headers()
+            .get("Accept-Encoding")
+            .flat_map(|accept| accept.split(","))
+            .map(|accept| accept.trim())
+            .any(|accept| accept == encoding)
+    }
+
+    fn already_encoded(response: &Response) -> bool {
+        response.headers().get("Content-Encoding").next().is_some()
+    }
+
+    fn set_body_and_encoding<'r, B: Read + 'r>(
+        response: &mut Response<'r>,
+        body: B,
+        encoding: Encoding,
+    ) {
+        response.set_header(ContentEncoding(vec![encoding]));
+        response.set_streamed_body(body);
+    }
+
+    fn skip_encoding(
+        content_type: &Option<rocket::http::ContentType>,
+        content_type_top: &Option<&UncasedStr>,
```

Comment:
This shouldn't take `content_type_top` separately, since it can be derived from `content_type`.

Reply:
I know it, but I put it as parameter because the function that call this function has to use `content_type_top` anyway some lines below, and the `content_type_top` extraction has to be done twice if it is not received as parameter. Should I remove the parameter anyway?

Reply:
You can leave it for now and I'll look again after the other changes.